doc/likwid-features.1
.TH LIKWID-FEATURES 1 <DATE> likwid\-<VERSION>
.SH NAME
likwid-features \- print and manipulate cpu features like hardware prefetchers (Intel only)
.SH SYNOPSIS
.B likwid-features
.RB [ \-vhal ]
.RB [ \-c
.IR cpus ]
.RB [ \-e
.IR taglist ]
.RB [ \-d
.IR taglist ]
.SH DESCRIPTION
.B likwid-features
is a command line application to print the flags in the model
specific register (MSR) MSR_IA32_MISC_ENABLE on Intel x86 processors. On Core2 and later processors
it can be used to toggle the hardware prefetch flags. It does not work on AMD processors.
For a documentation what flags are supported on which processor refer to the Intel
Software Developer's Manual Volume 3B, Table B.2 and https://software.intel.com/en-us/articles/disclosure-of-hw-prefetcher-control-on-some-intel-processors. The MSR are set individually for every hardware thread. Will be deprecated in 5.4.
The following hardware prefetchers can be toggled:
.IP \[bu]
.B HW_PREFETCHER:
Hardware prefetcher.
.IP \[bu]
.B CL_PREFETCHER:
Adjacent cache line prefetcher.
.IP \[bu]
.B DCU_PREFETCHER:
When the DCU prefetcher detects multiple loads from the same line done within a
time limit, the DCU prefetcher assumes the next line will be required. The next
line is prefetched in to the L1 data cache from memory or L2.
.IP \[bu]
.B IP_PREFETCHER:
The IP prefetcher is an L1 data cache prefetcher. The IP prefetcher looks for
sequential load history to determine whether to prefetch the next expected data
into the L1 cache from memory or L2.

.SH OPTIONS
.TP
.B \-\^v
prints version information to standard output, then exits.
.TP
.B \-\^h
prints a help message to standard output, then exits.
.TP
.B \-\^a
List out the names of all detected features
.TP
.B \-\^l
Print the state of all features for the given CPUs
.TP
.B \-\^c " cpus"
set on which hardware threads the MSR should be read and written. Syntax according to
.B likwid-pin(1)
.TP
.B \-\^d " HW_PREFETCHER | CL_PREFETCHER | DCU_PREFETCHER | IP_PREFETCHER"
specify which prefetcher should be disabled. Argument can be a comma-separated list.
.TP
.B \-\^e " HW_PREFETCHER | CL_PREFETCHER | DCU_PREFETCHER | IP_PREFETCHER"
specify which prefetcher should be enabled. Argument can be a comma-separated list.

.SH AUTHOR
Written by Thomas Gruber <thomas.roehl@googlemail.com>.
.SH BUGS
Report Bugs on <https://github.com/RRZE-HPC/likwidissues>.
.SH "SEE ALSO"
likwid-pin(1), likwid-topology(1), likwid-perfctr(1)


CHANGELOG
# Changelog 5.4.0
- Support for Intel Granite Rapids (core, energy, uncore)
- Support for Intel Sierra Forrest (core, energy, uncore)
- Support for AMD Bergamo (core, energy, uncore)
- Support for Nvidia Grace (core, uncore)
- Fix: AMD Zen4 DataFabric units
- Fix: Multi-socket RAPL measurements on Sapphire Rapids
- Fix: Energy unit for RAPL DRAM domain for SPR, GNR and SRF
- Fix: Discovery mechanism workaround for UPI and M3UPI units of SPR
- Fix: Intel Westmere Uncore with perf_event backend
- Fix: Fujitsu A64FX (more counters, fixed topology, ...)
- Container bridge to use LIKWID inside container
- Sysfeatures interface reworked with support for various architectures and libraries
- Fix build for NVMON for CUDA 12.6+
- likwid-mpirun: SLURM pinning with cpu_mask feature
- Update of internal hwloc (2.11.2) and Lua (5.4.7) version


# Changelog 5.3.0
- Support for Intel SapphireRapids (Core, Uncore, RAPL)
- Support for AMD Zen4 (Core, Uncore, RAPL)
- Support for Apple M1
- Support for AMD GPUs (MarkerAPI, F90 interface)
- Support for AWS Graviton3 (ARM Neoverse V1)
- Support for HiSilicon TSV110
- Fix of F90 interface installation
- Support for extended umasks in ICX and SPR
- Units for metrics in performance groups
- Library calls to get meta information (version, supported features, etc.)
- Some fixes for direct access mode
- Some fixes for X86 RDPMC detection
- Update of internal hwloc (2.9.3) and Lua (5.4.6) version
- New experimental sysfeatures module

# Changelog 5.2.2
- Add mutex to pinning library
- Fix pin string parsing in pinning library
- Make SBIN path configurable in build system
- Add PKGBUILD for ArchLinux package builds
- Remove accessDaemon double-fork in systemd environements
- Group updates for L2/L3 (mainly AMD Zen)
- Fix multi-initialization in MarkerAPI
- Add energy event scaling for Fujitsu A64FX
- Nvmon: Use Cupti error string to get better warning/error messages
- Nvmon: Store events internally to re-use event strings in stopCounters
- AccessLayer: Catch SIGCHLD to stop sending requests to accessDaemon if it was killed
- likwid-genTopoCfg: Update writing and reading of topology file
- Add INST_RETIRED_NOP event for Intel Icelake (desktop & server)
- Removed some memory leaks
- Improved checks for RDPMC availability
- Add TOPDOWN_SLOTS for perf_event
- Fix for systems with CPU sockets without hwthreads (A64FX FX1000)
- Fix if HOME environment variable is not set (systemd)
- Reader function for perf_event_paranoid in Lua to get state early
- likwid-mpirun: Sanitize np and ppn values to avoid crashes


# Changelog 5.2.1
- Add support for Intel Rocketlake and AMD Zen3 variant (Family 19, Model 0x50)
- Fix for perf_event multiplexing (important!)
- Fix for potential deadlock in MarkerAPI (thx @jenny-cheung)
- Build and runtime fixes for Nvidia GPU backend, updates for CUDA test codes
- peakflops kernel for ARMv8
- Updates for AMD Zen1/2/3 event lists and groups
- Support spaces in MarkerAPI region tags (thx @jrmadsen)
- Use 'online' cpulist instead of 'present'
- Switch CI from Travis-CI to NHR@FAU Cx services
- Document -reset and -ureset for likwid-setFrequencies
- Reset cpuset in unpinned runs
- Remove destructor in frequency module
- Check PID if given through --perfpid
- Intel Icelake: OFFCORE_RESPONSE events
- AccessDaemon: Check PCI init state before using it
- likwid-mpirun: Set mpi type for SLURM automatically
- likwid-mpirun: Fix for skip mask for OpenMPI


# Changelog 5.2.0
- Support for AMD Zen3 (Core + Uncore)
- Support for Intel IcelakeSP (Core + Uncore)
- New affinity code
- Fix for Ivybridge uncore code
- Bypass accessdaemon by using rdpmc instruction on x86_64
- Introduce notion of CPU die in topology module
- Use CPU dies for socket-lock for Intel CascadelakeAP
- Add environment variable LIKWID_IGNORE_CPUSET to break out of current CPUset
- Fixes for affinity module CPUlist sorting
- Build against system-installed hwloc
- Update for Intel SkylakeX/CascadelakeX L3 group
- Rename DataFabric events for all generations of AMD Zen
- Add static cache configuration for Fujitsu A64FX
- Add multiplexing checks for perf_event backend
- Fix for table width of likwid-topology afteradding CPU die column
- Adding RasPi 4 with 32 bit OS as ARMv7
- Add default groups for Intel Icelake desktop
- Fix for likwid-setFrequencies to not apply minFreq when setting governor
- likwid-powermeter: Fix hwthread selection when run with -p
- likwid-setFrequencies: Get measured base frequency if register is not readable
- CLOCK group for all AMD Zen
- Fixes in Nvidia GPU support in NvMarkerAPI and topology module

# Changelog 5.1.1
- Support for Intel Cometlake desktop (Core + Uncore)
- Fix for topology module of Fujitsu A64FX
- Fix for Intel Skylake SP in SNC mode
- Fix for likwid-perfscope
- Fix for CLI argument parsing
- Updated group and data file checkers
- Vector sum benchmark in SVE
- FP_PIPE group for Fujitsu A64FX
- Maximal number of CLI arguments configurable in config.mk (currently 16384)

# Changelog 5.1.0
- Support for Intel Icelake desktop (Core + Uncore)
- Support for Intel Icelake server (Core only)
- Support for Intel Tigerlake desktop (Core only)
- Support for Nvidia GPUs with compute capability >= 7.0 (CUpti Profiling API)
- Support for Fujitsu A64FX (Core) including SVE assembly benchmarks
- Support for ARM Neoverse N1 (AWS Graviton 2)
- Support for AMD Zen3 (Core + Uncore but without any events)
- Check for Intel HWP
- Fix for TID filter of Skylake SP LLC filter0 register
- Fix for Lua 5.1
- Fix for likwid-mpirun skip masks
- Fortran90 interface for NvMarkerAPI
- CPU_is_online check to filter non-usable CPU cores

# Changelog 5.0.2
- Fix memory leak in calc_metric()
- New peakflops benchmarks in likwid-bench
- Fix for NUMA domain handling properly
- Improvements for perf_event backend
- Fix for perfctr and powermeter with perf_event backend
- Fix for likwid-mpirun for SLURM with cpusets
- Fix for likwid-setFrequencies in cpusets

src/applications/likwid-features.lua
#!<INSTALLED_BINPREFIX>/likwid-lua
--[[
 * =======================================================================================
 *
 *      Filename:  likwid-features.lua
 *
 *      Description:  A application to retrieve and manipulate CPU features.
 *
 *      Version:   5.4.0
 *      Released:  15.11.2024
 *
 *      Author:   Thomas Gruber (tr), thomas.roehl@googlemail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2024 RRZE, University Erlangen-Nuremberg
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
]]
package.path = '<INSTALLED_PREFIX>/share/lua/?.lua;' .. package.path

local likwid = require("likwid")
local cpuinfo = likwid.getCpuInfo()

print_stdout = print
print_stderr = function(...) for k,v in pairs({...}) do io.stderr:write(v .. "\n") end end

GENERAL_FEATURES = {"HW_PREFETCHER", "CL_PREFETCHER", "DCU_PREFETCHER", "IP_PREFETCHER"}
KNL_FEATURES = {"HW_PREFETCHER", "DCU_PREFETCHER"}
FEATURES = GENERAL_FEATURES
if cpuinfo["short_name"] == "knl" then
    FEATURES = KNL_FEATURES
end

function version()
    print_stdout(string.format("likwid-features -- Version %d.%d.%d (commit: %s)",likwid.version,likwid.release,likwid.minor,likwid.commit))
end

function usage()
    version()
    print_stdout("A tool list and modify the states of CPU features.\n")
    print_stdout("Options:")
    print_stdout("-h, --help\t\t Help message")
    print_stdout("-v, --version\t\t Version information")
    print_stdout("-a, --all\t\t List all available features")
    print_stdout("-l, --list\t\t List features and state for given hardware threads")
    print_stdout("-c, --cpus <list>\t Perform operations on given hardware threads")
    print_stdout("-e, --enable <list>\t List of features that should be enabled")
    print_stdout("-d, --disable <list>\t List of features that should be disabled")
    print_stdout()
    print_stdout("Currently modifiable features:")
    print_stdout(table.concat(FEATURES, ", "))
end

if #arg == 0 then
    usage()
    os.exit(0)
end

listFeatures = false
num_cpus = 0
cpulist = {}
enableList = {}
disableList = {}
skipList = {}

for opt,arg in likwid.getopt(arg, {"h","v","l","c:","e:","d:","a","help","version","list", "enable:", "disable:","all", "cpus:"}) do
    if (type(arg) == "string") then
        local s,e = arg:find("-");
        if s == 1 then
            print_stderr(string.format("Argument %s to option -%s starts with invalid character -.", arg, opt))
            print_stderr("Did you forget an argument to an option?")
            os.exit(1)
        end
    end
    if opt == "h" or opt == "help" then
        usage()
        os.exit(0)
    elseif opt == "v" or opt == "version" then
        version()
        os.exit(0)
    elseif opt == "c" or opt == "cpus"then
        num_cpus, cpulist = likwid.cpustr_to_cpulist(arg)
    elseif opt == "l" or opt == "list" then
        listFeatures = true
    elseif opt == "a" or opt == "all" then
        if cpuinfo["isIntel"] == 0 then
            print_stdout("INFO: Manipulation of CPU features is only available on Intel platforms")
        end
        print_stdout("Available features:")
        for i=0,likwid.tablelength(likwid.cpuFeatures)-1 do
            local found = false
            for j, f in pairs(FEATURES) do
                if likwid.cpuFeatures[i] == f then
                    found = true
                    break
                end
            end
            if found then
                print_stdout(string.format("\t%s*",likwid.cpuFeatures[i]))
            else
                print_stdout(string.format("\t%s",likwid.cpuFeatures[i]))
            end
        end
        print_stdout("Modifiable features are marked with *")
        os.exit(0)
    elseif opt == "e" or opt == "enable" then
        local tmp = likwid.stringsplit(arg, ",")
        for i, f in pairs(tmp) do
            for i=0,likwid.tablelength(likwid.cpuFeatures)-1 do
                if likwid.cpuFeatures[i] == f then
                    table.insert(enableList, i)
                end
            end
        end
    elseif opt == "d" or opt == "disable" then
        local tmp = likwid.stringsplit(arg, ",")
        for i, f in pairs(tmp) do
            for i=0,likwid.tablelength(likwid.cpuFeatures)-1 do
                if likwid.cpuFeatures[i] == f then
                    table.insert(disableList, i)
                end
            end
        end
    elseif opt == "?" then
        print_stderr("Invalid commandline option -"..arg)
        os.exit(1)
    elseif opt == "!" then
        print_stderr("Option requires an argument")
        os.exit(1)
    end
end

if cpuinfo["isIntel"] == 0 then
    print_stdout("INFO: Manipulation of CPU features is only available on Intel platforms")
    os.exit(0)
end

likwid.initCpuFeatures()

if listFeatures and #cpulist > 0 then
    if likwid.getCpuFeatures(c, 0) >= 0 then
        local str = "Feature"..string.rep(" ",string.len("BRANCH_TRACE_STORAGE")-string.len("Feature")+2)
        for j, c in pairs(cpulist) do
            str = str..string.format("HWThread %d\t",c)
        end
        print_stdout(str)
        str = ""
        for i=0,likwid.tablelength(likwid.cpuFeatures)-1 do
            str = likwid.cpuFeatures[i]..string.rep(" ",string.len("BRANCH_TRACE_STORAGE")-string.len(likwid.cpuFeatures[i])+2)
            for j, c in pairs(cpulist) do
                if (likwid.getCpuFeatures(c, i) == 1) then
                    str = str .. "on\t"
                else
                    str = str .. "off\t"
                end
            end
            print_stdout(str)
        end
    end
    os.exit(1)
elseif #cpulist == 0 then
    print_stderr("Need HWThread to list current feature state")
    os.exit(1)
end

if #enableList > 0 and #disableList > 0 then
    for i,e in pairs(enableList) do
        for j, d in pairs(disableList) do
            if (e == d) then
                print_stderr(string.format("Feature %s is in enable and disable list, doing nothing for feature", e))
                table.insert(skipList, e)
            end
        end
    end
    for i, s in pairs(skipList) do
        for j, e in pairs(enableList) do
            if (s == e) then table.remove(enableList, j) end
        end
        for j, e in pairs(disableList) do
            if (s == e) then table.remove(disableList, j) end
        end
    end
end

if #enableList > 0 then
    for i, c in pairs(cpulist) do
        for j, f in pairs(enableList) do
            local ret = likwid.enableCpuFeatures(c, f, 1)
            if ret == 0 then
                print_stdout(string.format("Enabled %s for HWThread %d", likwid.cpuFeatures[f], c))
            else
                print_stdout(string.format("Failed %s for HWThread %d", likwid.cpuFeatures[f], c))
            end
        end
    end
end
if #disableList > 0 then
    for i, c in pairs(cpulist) do
        for j, f in pairs(disableList) do
            local ret = likwid.disableCpuFeatures(c, f, 1)
            if ret == 0 then
                print_stdout(string.format("Disabled %s for HWThread %d", likwid.cpuFeatures[f], c))
            else
                print_stdout(string.format("Failed %s for HWThread %d", likwid.cpuFeatures[f], c))
            end
        end
    end
end
